source/fuzz/fuzzer_pass_add_composite_inserts.cpp

```diff
@@ -0,0 +1,256 @@
+          num_of_components =
+              GetNumberOfComponents(GetIRContext(), current_node_type_id);
+          one_selected_index = GetFuzzerContext()->GetRandomIndexForAccessChain(
+              num_of_components);
+
+          // Construct a final index by appending the current index.
+          index_to_replace.push_back(one_selected_index);
+          current_node_type_id = fuzzerutil::WalkOneCompositeTypeIndex(
+              GetIRContext(), current_node_type_id, one_selected_index);
+
+          // If the component is not a composite or if we decide not to go
+          // deeper, then end the iteration.
+          if (!fuzzerutil::IsCompositeType(
+                  GetIRContext()->get_type_mgr()->GetType(
+                      current_node_type_id)) ||
+              !GetFuzzerContext()->ChoosePercentage(
+                  GetFuzzerContext()
+                      ->GetChanceOfGoingDeeperToInsertInComposite())) {
+            reached_end_node = true;
+          }
+        }
+
+        // Look for available objects that have the type id
+        // |current_node_type_id| and can be inserted.
+        std::vector<opt::Instruction*> available_objects =
+            FindAvailableInstructions(
+                function, block, instruction_iterator,
+                [instruction_descriptor, current_node_type_id](
+                    opt::IRContext* ir_context,
+                    opt::Instruction* instruction) -> bool {
+                  if (instruction->type_id() != current_node_type_id) {
+                    return false;
+                  }
+                  return fuzzerutil::IdIsAvailableBeforeInstruction(
+                      ir_context,
+                      FindInstruction(instruction_descriptor, ir_context),
+                      instruction->result_id());
+                });
+
+        // If there are no objects of the specific type available, create a zero
+        // constant of this type.
+        uint32_t available_object_id;
+        if (available_objects.empty()) {
+          available_object_id =
+              FindOrCreateZeroConstant(current_node_type_id, false);
+        } else {
+          available_object_id =
+              available_objects[GetFuzzerContext()->RandomIndex(
+                                    available_objects)]
+                  ->result_id();
+        }
+
+        auto new_result_id = GetFuzzerContext()->GetFreshId();
+
+        // Insert an OpCompositeInsert instruction which copies
+        // |available_composite| and in the copy inserts the object
+        // of type |available_object_id| at index |index_to_replace|.
+        TransformationCompositeInsert transformation =
+            TransformationCompositeInsert(instruction_descriptor, new_result_id,
+                                          available_composite->result_id(),
+                                          available_object_id,
+                                          std::move(index_to_replace));
```

Comment:
You have moved `index_to_replace` but then you use it below. I don't think you want to move it here. I guess copying is fine.